File: 3.x.x/crates-utopia/lock/src/lock.rs
use crate::error::Contention;

/// PHP `Utopia\Lock\Lock`.
///
/// `timeout` is seconds. Negative waits forever; `0.0` does not wait.
pub trait Lock: Send + Sync {
    fn acquire(&self, timeout: f64) -> bool;
    fn try_acquire(&self) -> bool;
    fn release(&self);

    /// PHP `withLock($callback, $timeout = 0.0)`.
    fn with_lock<R, F: FnOnce() -> R>(&self, callback: F, timeout: f64) -> Result<R, Contention> {
        if !self.acquire(timeout) {
            return Err(self.contention());
        }
        let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(callback));
        self.release();
        match result {
            Ok(value) => Ok(value),
            Err(payload) => std::panic::resume_unwind(payload),
        }
    }

    fn contention(&self) -> Contention {
        Contention::new("Failed to acquire lock within timeout")
    }
}

File: packages/lock/src/Lock.php
<?php

declare(strict_types=1);

namespace Utopia\Lock;

interface Lock
{
    /**
     * Block up to $timeout seconds trying to acquire the lock.
     *
     * Passing 0.0 means do not wait; a negative value means wait forever.
     *
     * @return bool true if the lock was acquired
     */
    public function acquire(float $timeout = 0.0): bool;

    /**
     * Non-blocking acquire attempt.
     *
     * @return bool true if the lock was acquired
     */
    public function tryAcquire(): bool;

    /**
     * Release the lock. Safe to call when the lock is not held.
     */
    public function release(): void;

    /**
     * Acquire the lock, run $callback, then release the lock.
     *
     * The lock is always released, including when $callback throws.
     *
     * @template T
     *
     * @param  callable(): T  $callback
     * @return T
     *
     * @throws \Utopia\Lock\Exception\Contention when acquire times out
     */
    public function withLock(callable $callback, float $timeout = 0.0): mixed;
}
